### PR TITLE
track transaction changes

### PR DESCRIPTION
**Pull request by drewbailey**
This commit wraps memdb.DB with a changeTrackerDB, which is a thin
wrapper around memdb.DB which enables go-memdb's TrackChanges on all write
transactions. When a transaction is committed the changes are sent to
an eventPublisher which will be used to create and emit change events.

### CODE REVIEW

File: nomad/event/event.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package event
```
Comment (drewbailey) [inline nomad/event/event.go:1]: stub package for now

File: nomad/state/node_events.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
+package state
+
+import (
```
Comment (drewbailey) [inline nomad/state/node_events.go:3]: this file and its corresponding tests are meant to exercise the transaction changes but will be changing greatly in future commits

File: nomad/state/state_store.go [hunk 3]
```diff
@@ -13,12 +13,13 @@ import (
 	"github.com/pkg/errors"
 
 	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/event"
```
Comment (drewbailey) [inline nomad/state/state_store.go:16]: bulk of the changes are in state_store.go, commenting to force github to open the diff

File: nomad/state/state_store.go [hunk 4]
```diff
@@ -4087,19 +4108,6 @@ func (s *StateStore) Index(name string) (uint64, error) {
 	return out.(*IndexEntry).Value, nil
 }
 
-// RemoveIndex is a helper method to remove an index for testing purposes
-func (s *StateStore) RemoveIndex(name string) error {
```
Comment (drewbailey) [inline nomad/state/state_store.go:4091]: unreachable code

File: nomad/state/node_events.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	var events []event.Event
+
+	var nodeChanges map[string]*memdb.Change
+
+	markNode := func(node string, nodeChange *memdb.Change) {
+		if nodeChanges == nil {
+			nodeChanges = make(map[string]*memdb.Change)
+		}
+		ch := nodeChanges[node]
+		if ch == nil {
```
Comment (langmartin) [inline nomad/state/node_events.go:33]: It looks like we're throwing away the second change to a node, why don't we want an event for that too?

File: nomad/state/state_store.go [hunk 6]
```diff
@@ -833,28 +856,26 @@ func (s *StateStore) UpdateNodeStatus(index uint64, nodeID, status string, updat
...
 	// Insert the node
 	if err := txn.Insert("nodes", copyNode); err != nil {
 		return fmt.Errorf("node update failed: %v", err)
 	}
-	if err := txn.Insert("index", &IndexEntry{"nodes", index}); err != nil {
+	if err := txn.Insert("index", &IndexEntry{"nodes", txn.Index}); err != nil {
 		return fmt.Errorf("index update failed: %v", err)
 	}
-
-	txn.Commit()
```
Comment (langmartin) [inline nomad/state/state_store.go:851]: why doesn't this commit here anymore?

Comment (drewbailey) [reply nomad/state/state_store.go:851]: sketchy github diff, UpdateNodeStatus initializes the transaction, calls updateNodeStatusTxn to do the logic, and returns an error, the transactions committed in UpdateNodeStatus

File: nomad/state/node_events.go (hunk 5 above)
Comment (drewbailey) [reply nomad/state/node_events.go:33]: We do, I'll need to go back and edit this, most of this file is likely going to change but wanted to exercise the transaction changes bits with what node event sourcing may look like

**Review by langmartin — APPROVED** [review]

File: nomad/state/node_events.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+func (s *StateStore) AllocsByNodeTx(tx ReadTxn, node string) ([]*structs.Allocation, error) {
+	iter, err := tx.Get("allocs", "node_prefix", node)
+	if err != nil {
+		return nil, err
+	}
+
+	var out []*structs.Allocation
+	for {
+		raw := iter.Next()
```
Comment (shoenig) [inline nomad/state/node_events.go:168]: these bits can be the for condition

File: nomad/state/state_changes.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+func (n *noOpPublisher) Publish(events []event.Event)            {}
+func noOpProcessChanges(ReadTxn, Changes) ([]event.Event, error) { return []event.Event{}, nil }
+
+// Txn exists to maintain backwards compatibility with memdb.DB.Txn. Preexisting
+// code may use it to create a read-only transaction, but it will panic if called
+// with write=true.
+//
+// Deprecated: use either ReadTxn, or WriteTxn.
+func (c *changeTrackerDB) Txn(write bool) *txn {
```
Comment (shoenig) [inline nomad/state/state_changes.go:50]: If we own all the code that could access this, lets just remove it

File: nomad/state/state_store.go [hunk 9]
```diff
@@ -76,10 +77,14 @@ func NewStateStore(config *StateStoreConfig) (*StateStore, error) {
...
 		logger:    config.Logger.Named("state_store"),
-		db:        db,
 		config:    config,
 		abandonCh: make(chan struct{}),
 	}
+	s.db = &changeTrackerDB{
+		db:             db,
+		publisher:      event.NewPublisher(),
+		processChanges: processDBChanges,
+	}
```
Comment (shoenig) [inline nomad/state/state_store.go:87]: with this much fanciness required to make a `StateStore` it's probably time to use a constructor making sure the necessary things get passed in and set

File: nomad/event/event.go [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+type Event struct {
+	Topic   string
+	Key     string
+	Index   uint64
+	Payload interface{}
+}
+
+type EventPublisher struct{}
```
Comment (shoenig) [inline nomad/event/event.go:10]: looking forward to this!

**Review by shoenig — APPROVED** [review]
There's some nil pointer check I think is missing, but otherwise I think this LGTM. 

For other large features in the past we've created long lived feature branches into which we review and merge the smaller PRs. We might want to do that here as well, since this is largely stubs still. The downside of course is the joy of either regular rebasing with master, or resolving a big merge at the very end - more of a @schmichael question

**Comment by drewbailey** [comment]
> For other large features in the past we've created long lived feature branches into which we review and merge the smaller PRs. We might want to do that here as well, since this is largely stubs still. The downside of course is the joy of either regular rebasing with master, or resolving a big merge at the very end - more of a @schmichael question

I'm cool with having a long lived feature branch with the caveat being I'd like to at least merge in the state_store changes now since it affects every single function signature, rebasing and keeping it up to date will be tiresome. I can back out the node event changes and everything else should be able to live in a feature branch.

I'll go ahead and back out the node event stuff from the PR

**Comment by drewbailey** [comment]
@shoenig removed `node_events` from this PR, also dropped Txn for ReadTxn

File: nomad/state/state_changes.go (hunk 8 above)
Comment (drewbailey) [reply nomad/state/state_changes.go]: updated state_store to use ReadTxn()

**Pull request merged**